src/snowplow/snowplow.lua

```diff
@@ -17,54 +17,52 @@
 local validate = require("validate")
 local tracker = require("tracker")
+local set = require("lib.set")
 
 local snowplow = {}
+local protocols = set.new_set({ "http", "https" })
 
 -- --------------------------------------------------------------
 -- Static methods
 
--- Builds our new tracker using the supplied URI.
--- @param uri string: The full URI to the Snowplow collector
--- @return tracker table: The new tracker
-local function init_tracker(uri)
-  return tracker.new_tracker(uri)
-end
-
--- Helper to generate the collector url from a collector host name.
--- Example: as_collector_url("snplow.myshop.com") => "http://snplow.myshop.com/i"
--- @param host string: The host name of the collector
+-- Helper to generate the collector url from a collector uri.
+-- Example: as_collector_url("snplow.myshop.com") => "https://snplow.myshop.com/i"
+-- @param endpoint string: The endpoint name of the collector
+-- @param protocol string: The protocol to use (defaults to "https")
 -- @return string: The full collector URI
-local function as_collector_uri(host)
-  return "http://" .. host .. "/i"
-end
+local function as_collector_uri(endpoint, protocol)
+  if not endpoint:find("/i$") then
+    endpoint = endpoint .. "/i"
+  end
 
--- Helper to generate the collector url from a CloudFront distribution subdomain.
--- Example: collector_uri_from_cf("f3f77d9def5") => "http://f3f77d9def5.cloudfront.net/i"
--- @param cf_subdomain string The CloudFront subdomain on which the collector's distribution is hosted
--- @return string: The full collector URI
-local function collector_uri_from_cf(cf_subdomain)
-  return as_collector_uri(cf_subdomain .. ".cloudfront.net")
-end
+  if protocol ~= nil and endpoint:find("^https?://") and protocols:contains(protocol) then
+    print(
+      "WARN: Ignoring protocol argument '"
+        .. protocol
+        .. "', protocol already specified in endpoint '"
+        .. endpoint:sub(1, endpoint:find("/i$") - 1) -- Don't include '/i' in the warning
+        .. "'"
+    )
+    return endpoint
+  end
 
--- --------------------------------------------------------------
--- Factories to create a tracker
+  protocol = protocol or "https"
 
--- Create a new Snowplow tracker talking to a URI-based collector on the given host.
--- @param host string: The host (i.e. full domain) on which the collector is running
--- @return tracker table: The new tracker
-function snowplow.new_tracker_for_uri(host)
-  validate.is_non_empty_string("host", host)
-  local uri = as_collector_uri(host)
-  return init_tracker(uri)
+  return protocol .. "://" .. endpoint
 end
 
--- Create a new Snowplow tracker talking to a CloudFront-based collector on the given subdomain.
--- @param cf_subdomain string: The CloudFront subdomain on which the collector is running
+-- Builds our new tracker using the supplied URI.
+-- @param endpoint string: The full endpoint to the Snowplow collector
+-- @param protocol string: An optional protocol to use (defaults to "https")
```

Comment:
This is a bit unusual and non-intuitive to me. I would expect that the protocol can be part of the endpoint URI. That is the approach that we take in most other trackers and as we are trying to make their APIs more consistent, it would be great to adopt it here. I think it'd be better if the behaviour was:

- if the endpoint contains a protocol at the start of the URI, use that protocol
- if it doesn't contain a protocol, use the default https as now

No need for the explicit protocol attribute. [Here](https://github.com/snowplow/snowplow-android-tracker/blob/master/snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/configuration/NetworkConfiguration.java#L77-L101) is an example how the endpoint is parsed on the Android tracker.

Reply:
I understand that it is already allowed for the endpoint URI to contain the protocol but I think that the optional protocol attribute is unnecessary and only adds confusion. It also unnecessarily complicates the code as now it has two places to look for protocols. If you think it is useful, go for it, but I would vote for simplicity and avoiding confusion (have only one way to do a thing).

Reply:
https://github.com/snowplow/snowplow-lua-tracker/pull/31#issuecomment-1066749113